### PR TITLE
Xhr data

### LINKED ISSUE #8015: Allow requesting data URLs from XMLHttpRequest
Step 5 of https://xhr.spec.whatwg.org/#dom-xmlhttprequest-send explicitly sets the `same-origin data-URL flag` to true. We have a TODO about this in `CORSRequest::maybe_new`; I propose creating a new enum representing this flag, and passing it as an argument to `CORSRequest::maybe_new`. This should cause the output of `./mach test-wpt tests/wpt/web-platform-tests/XMLHttpRequest/data-uri.htm` to improve.

Note: this is reserved for a potential Outreachy contributor who I am in contact with.

Code: `components/script/cors.rs`, `components/script/dom/xmlhttprequest.rs`

### PR DESCRIPTION
**Pull request by dagnir**
Builds on existing work by @emosenkis.  Fixes #8015.

---

This change is [Review on Reviewable](https://reviewable.io/reviews/servo/servo/9972)

### CODE REVIEW

**Comment by KiChjang** [comment]
Why is `tests/wpt/web-platform-tests/XMLHttpRequest/data-uri.htm` modified?

**Comment by dagnir** [comment]
Based on the discussion on #8807 and looking at the spec, when the response is an error the status message is empty and not 'OK' like the test expects.
https://fetch.spec.whatwg.org/#concept-network-error

As for the comparison for 'Content-Type', servo seems to return the string 'text/html; charset=utf-8', with the space, which doesn't quite match the original expected value.

File: tests/wpt/web-platform-tests/XMLHttpRequest/data-uri.htm [hunk 1]
```diff
@@ -23,14 +23,14 @@
         if (method.toUpperCase() !== 'GET') {
           assert_equals(client.status, 0);
           assert_equals(client.responseText, '');
-          assert_equals(client.statusText, 'OK');
+          assert_equals(client.statusText, '');
```
Comment (KiChjang) [inline tests/wpt/web-platform-tests/XMLHttpRequest/data-uri.htm]: Specifically, why is this modified from OK to an empty string?

File: tests/wpt/web-platform-tests/XMLHttpRequest/data-uri.htm [hunk 2]
```diff
@@ -43,7 +43,7 @@
   do_test('GET', "data:text/plain,Hello, World!");
   do_test('GET', "data:text/plain;base64,SGVsbG8sIFdvcmxkIQ==", undefined, " (base64)");
   do_test('GET', "data:text/html,Hello, World!", 'text/html');
-  do_test('GET', "data:text/html;charset=UTF-8,Hello, World!", 'text/html;charset=UTF-8');
+  do_test('GET', "data:text/html;charset=UTF-8,Hello, World!", 'text/html;charset=utf-8');
```
Comment (KiChjang) [inline tests/wpt/web-platform-tests/XMLHttpRequest/data-uri.htm]: Why is this lower-cased?

**Comment by KiChjang** [comment]
My concern here is whether Firefox and other browsers pass this test as-is without your changes.

**Comment by dagnir** [comment]
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[tests/wpt/web-platform-tests/XMLHttpRequest/data-uri.htm, line 26 \[r2\]](https://reviewable.io:443/reviews/servo/servo/9972#-KCbjIhVEUX4goqoEiL4-r2-26:-KCcF9IDI3rRGsDtw5F0:-908269660) ([raw file](https://github.com/servo/servo/blob/4fedbfbc1b91e4fa9e3745c31844b53920294063/tests/wpt/web-platform-tests/XMLHttpRequest/data-uri.htm#L26)):*
Spec says status text is empty when status == 0

---

*[tests/wpt/web-platform-tests/XMLHttpRequest/data-uri.htm, line 46 \[r2\]](https://reviewable.io:443/reviews/servo/servo/9972#-KCbjIhVEUX4goqoEiL4-r2-46:-KCcErrbFblYK_IX-bLT:-255230216) ([raw file](https://github.com/servo/servo/blob/4fedbfbc1b91e4fa9e3745c31844b53920294063/tests/wpt/web-platform-tests/XMLHttpRequest/data-uri.htm#L46)):*
I was looking at this https://www.w3.org/Protocols/rfc2616/rfc2616-sec3.html#sec3.7, but I seem to have missed the part about parameter *values* being case sensitive.  Do you think it matters for this particular case (utf-8 vs UTF-8)?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9972)*

**Comment by dagnir** [comment]
Regarding opening a PR with https://github.com/w3c/web-platform-tests, should I revert the changes here and wait for that PR to land before proceeding with this one?

**Comment by KiChjang** [comment]
Yes, if the tests are wrong, open a PR at w3c/web-platform-tests and wait until that PR is merged.

**Comment by dagnir** [comment]
Thanks, the PR is here: https://github.com/w3c/web-platform-tests/pull/2667.  I will update once it's merged and the servo copy is synced.

**Comment by jdm** [comment]
Still waiting on an update of our local WPT clone at this point.

**Comment by jdm** [comment]
It merged!

**Comment by KiChjang** [comment]
@bors-servo r+

Thanks!
